### DISCUSSION TITLE
Discord Screenshots at Motions

### DISCUSSION THREAD

**Discussion by jay-oswald** [briis/unifiprotect#273] [Show and tell]
I've setup an automation so that whenever a camera detects motion and nobody is home it sends me a discord message.

With a frame from when the motion was triggered, which camera triggered it as well as a timestamp and a link to view the live cameras.

![image](https://user-images.githubusercontent.com/16188552/127338575-b6bcc541-029f-4088-9b26-db6e0234611f.png)

If anybody else is interested let me know and I'll write up how to get it running. The very quick version is Home Assistant calls a node script running in its own container that does all the work.

It should be very useful, as a motion alert is pretty useless without seeing what may have caused it. Without the image I never really bothered to open the app it was to much work.

One thing that still needs to be fixed is getting the link to open up the protect app when I click on it on Discord, at the moment it just opens up my cloudkey in a browser, which works since my phone is VPN'ed into my local network, but would much rather it open the app. So if anybody knows how I'd love to fix it.

**Comment by skrilla0** [comment]
Hey @jay-oswald just discovered this post. given some of Unifi Protect's reliability issues with real time push notifications, would love to have a Discord integration. Can you please share?

**Reply by jay-oswald** [reply]
> Hey @jay-oswald just discovered this post. given some of Unifi Protect's reliability issues with real time push notifications, would love to have a Discord integration. Can you please share?

The integration isn't really easily sharable in itself, since its just all hard-coded values running on a node.js server, its not an addon or integration or anything.

But I'm running a separate node.js server on my Pi (using HA inside of its own docker container) in its own docker container and using HA to send a webhook to the node.js server do do the work, get the screenshot and send the notification. I've written it in typescript, so its a bit over-complicated for this, but I like things structured, and I figured I would  probably end up adding more custom webhook endpoints in the future. Below is the primary function that does all the work.

The url is something like `/protect/motion-alert/:cameraName` so 1 webhook per camera, so a bit annoying to set up, but I only have 1 camera for now, only planning on having about 4 total, would be a pain if you had much more. Im sure you could write it to check with your protect server to dynamically get the camera with the last motion event, but that's overkill for me, at least for now.

So all the script does is create a temporary file with the current timestamp and then saves the camera snapshot to it, which can be found at `http://${this.ip}/snap.jpeg` you do need to log into the cameras interface (not protect) to turn on snapshots, since the option to toggle it isn't in protect. Then it just messages discord with some text, timestamp, which camera etc as well as the snapshot.

I have the HA rules set to only use the webhook when there is nobody home, and its been working well, been using it ever since I made the original post.

Hopefully all that and the code below is enough for you to get it up and running on your own server.

The only thing thats really important/missing is my camera class, where it stores all the cameras and I get the details, like their IP by their name.


```export async function motionAlert(req: express.Request, res: express.Response) {
    const cameraName = req.params.cameraName;
    const camera = getCameraByName(cameraName);
    if (camera === null) {
        res.status(404).send('404');
        return;
    }


    const tempFileName = `${new Date().getTime()}.jpeg`;
    const tempFilePath = `/tmp/${tempFileName}`;
    debug(tempFilePath);
    const file = fs.createWriteStream(tempFilePath);

    const response = await axios({
        url: camera.snapshotUrl(),
        method: 'GET',
        responseType: 'stream'
    });

    response.data.pipe(file);

    await new Promise((resolve, reject) => {
        response.data.on('end', () => {
            resolve();
        })
        response.data.on('error', () => {
            reject();
        })
    })

    file.close();

    const channelId = 'YOUR_CHANNEL_ID_HERE';
    const guildId = 'YOUR_SERVER_ID_HERE';

    const channel = await discordClient.channels.cache.find(channel => channel.id === channelId);

   const discordAttachment = await new MessageAttachment(tempFilePath, tempFileName);

   const date = new Date();
   const currentTime = `${date.getHours()}:${date.getMinutes()}:${date.getSeconds()}`

    const messageEmbed = new MessageEmbed();
   messageEmbed.setTitle(`Motion Detected`)
    messageEmbed.setDescription("[Live View](https://cloudkey.local/protect/liveview/)")
    messageEmbed.attachFiles([tempFilePath])
    messageEmbed.setImage(`attachment://${tempFileName}`);
   messageEmbed.addFields(
       {name: "Camera", value: camera.name_pretty},
       {name: "Time", value: currentTime}
   );

    await (channel as TextChannel).send(messageEmbed);

    res.send(200);

}```
